pkg/common/cns-lib/volume/taskmanager.go

```diff
@@ -0,0 +1,355 @@
+				prop := types.PropertyChange{
+					Val: types.TaskInfo{
+						Task:  taskDetails.Reference,
+						State: types.TaskInfoStateError,
+						Error: &types.LocalizedMethodFault{
+							LocalizedMessage: err.Error(),
+						},
+					},
+				}
+				go t.handleTaskUpdate(ctx, prop, taskDetails.MarkedForRemoval)
+			}
+			time.Sleep(waitForUpdatesRetry)
+		}
+	}
+}
+
+// handleTaskUpdate is processes each task update in a separate goroutine
+func (t *TaskViewManagerImpl) handleTaskUpdate(ctx context.Context, prop types.PropertyChange, isMarkedForRemoval bool) {
+	log := logger.GetLogger(ctx)
+	log.Infof("handleTaskUpdate for prop: %+v", prop)
+	taskInfo := prop.Val.(types.TaskInfo)
+	if taskInfo.State == types.TaskInfoStateQueued || taskInfo.State == types.TaskInfoStateRunning {
+		log.Infof("task not completed yet. current state: %v", taskInfo.State)
+		return
+	}
+	result := Result{}
+	taskDetails := TaskDetails{
+		Reference:        taskInfo.Task,
+		MarkedForRemoval: isMarkedForRemoval,
+	}
+	requestorChan, err := t.getChannelFromTask(ctx, taskDetails)
+	if err != nil {
+		result.TaskInfo = nil
+		result.Err = err
+	} else if taskInfo.State == types.TaskInfoStateError {
+		log.Infof("taskInfo.Error: %+v", taskInfo.Error)
+		result.TaskInfo = nil
+		result.Err = errors.New(taskInfo.Error.LocalizedMessage)
+	} else {
+		result.TaskInfo = &taskInfo
+		result.Err = nil
+	}
+
+	requestorChan <- result
+}
+
+// AddTaskToListView adds task to listView and the internal map
+func (t *TaskViewManagerImpl) AddTaskToListView(ctx context.Context, taskMoRef types.ManagedObjectReference, ch chan Result) error {
+	log := logger.GetLogger(ctx)
+	log.Infof("AddTaskToListView for %+v", taskMoRef)
+	for t.getListView() == nil {
+		log.Infof("listview not created. waiting for 0.5 sec")
+		time.Sleep(time.Millisecond * 500)
+	}
+	listView := t.getListView()
+	log.Infof("listview -> %+v", listView)
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	err := listView.Add(ctx, []types.ManagedObjectReference{taskMoRef})
+	if err != nil {
+		return logger.LogNewErrorf(log, "failed to add task to ListView. error: %+v", err)
+	}
+	log.Infof("Added task %+v to listView", taskMoRef)
+	t.addTaskToMap(ctx, taskMoRef, ch)
+	return nil
+}
+
+func (t *TaskViewManagerImpl) addTaskToMap(ctx context.Context, taskMoRef types.ManagedObjectReference, ch chan Result) {
+	t.taskToChannelMap[TaskDetails{Reference: taskMoRef}] = ch
+	log := logger.GetLogger(ctx)
+	log.Infof("Added task %+v and channel to internal map", taskMoRef)
+}
+
+func (t *TaskViewManagerImpl) removeTaskFromMap(ctx context.Context, taskMoRef types.ManagedObjectReference) {
+	delete(t.taskToChannelMap, TaskDetails{Reference: taskMoRef})
+	log := logger.GetLogger(ctx)
+	log.Infof("Removed task %+v from internal map", taskMoRef)
+}
+
+func (t *TaskViewManagerImpl) getChannelFromTask(ctx context.Context, taskDetails TaskDetails) (chan Result, error) {
+	log := logger.GetLogger(ctx)
+	t.mu.Lock()
+	defer t.mu.Unlock()
```

Comment:
You aren't locking this when you are adding/deleting to the map, but locking it here. Could you explain this

Reply:
Add/Delete to map are also under lock. I was applying the lock in the methods that call add and delete. I'll do the same for this method to keep it consistent.